src/_zkapauthorizer/tests/test_storage_server.py

```diff
@@ -524,3 +547,173 @@ def test_mutable_share_sizes(
+        secrets,
+        test_and_write_vectors_for_shares,
+    ):
+        tw_vectors = {
+            k: v.for_call() for (k, v) in test_and_write_vectors_for_shares.items()
+        }
+        num_passes = get_required_new_passes_for_mutable_write(
+            self.pass_value,
+            dict.fromkeys(tw_vectors.keys(), 0),
+            tw_vectors,
+        )
+        valid_passes = make_passes(
+            self.signing_key,
+            slot_testv_and_readv_and_writev_message(storage_index),
+            list(RandomToken.create() for i in range(num_passes)),
+        )
+
+        # Create an initial share to toy with.
+        test, read = self.storage_server.doRemoteCall(
+            "slot_testv_and_readv_and_writev",
+            (),
+            dict(
+                passes=valid_passes,
+                storage_index=storage_index,
+                secrets=secrets,
+                tw_vectors=tw_vectors,
+                r_vector=[],
+            ),
+        )
+
+        after_count = read_count(self.storage_server)
+        after_bucket = read_bucket(self.storage_server, num_passes)
+
+        self.expectThat(
+            after_count,
+            Equals(1),
+            "Unexpected histogram sum value",
+        )
+        self.assertThat(
+            after_bucket,
+            Equals(1),
+            "Unexpected histogram bucket value",
+        )
+
+    @given(
+        storage_index=storage_indexes(),
+        renew_secret=lease_renew_secrets(),
+        cancel_secret=lease_cancel_secrets(),
+        sharenums=sharenum_sets(),
+        size=sizes(),
+    )
+    def test_immutable_spending_metrics(
+        self, storage_index, renew_secret, cancel_secret, sharenums, size
+    ):
+        """
+        When ZKAPs are spent to call *allocate_buckets* the number of passes spent
+        is recorded as a metric.
+        """
+        num_passes = required_passes(
+            self.storage_server._pass_value, [size] * len(sharenums)
+        )
+        valid_passes = make_passes(
+            self.signing_key,
+            allocate_buckets_message(storage_index),
+            list(RandomToken.create() for i in range(num_passes)),
+        )
+
+        alreadygot, allocated = self.storage_server.doRemoteCall(
+            "allocate_buckets",
+            (),
+            dict(
+                passes=valid_passes,
+                storage_index=storage_index,
+                renew_secret=renew_secret,
+                cancel_secret=cancel_secret,
+                sharenums=sharenums,
+                allocated_size=size,
+                canary=LocalReferenceable(None),
+            ),
+        )
+
+        after_count = read_count(self.storage_server)
+        after_bucket = read_bucket(self.storage_server, num_passes)
+
+        self.expectThat(
+            after_count,
+            Equals(1),
+            "Unexpected histogram sum value",
+        )
+        self.assertThat(
+            after_bucket,
+            Equals(1),
```

Comment:
Should we be checking the other buckets are 0?

Reply:
They must be 0 because the sum is 1.  Or prometheus_client is broken. :)

Reply:
I guess that is true, but if it isn't too much work, I might be inclined to make the strong assertion anyway; I think the test would be slightly easier to reason about that way. (We could possibly even skip the sum assertion instead).

Reply:
That's reasonable.  I started to do this and realized I gave an incorrect explanation - `read_count` does return the sum, but it's the sum we calculate by looking at all of the buckets, not the sum prometheus_client calculates.  So the other buckets must be 0 or `read_count` (now `read_spending_success_histogram_total`) is broken.